### PR TITLE
Feature/logger

### PR DESCRIPTION
**Pull request by oklymchuk**

### CODE REVIEW

File: config/config.go [hunk 1]
```diff
@@ -2,3 +2,30 @@
...
+	"github.com/lvl484/user-manager/logger"
+	"github.com/spf13/viper"
+)
+
+// NewPostgresConfig returns pointer to PointerConfig with data read from viper.config.json
+func NewLoggerConfig(configName, configPath string) (*logger.LogConfig, error) {
+	v := viper.New()
+	v.AddConfigPath(configPath)
+	v.SetConfigName(configName)
+	v.SetEnvKeyReplacer(strings.NewReplacer(".", "_"))
```
Comment (crnopster) [inline config/config.go]: Would you call that for every new func for getting configs?

File: config/config.json [hunk 2]
```diff
@@ -0,0 +1,9 @@
+{
+    "logger": {
+        "Host": "localhost",
+        "port": 12201,
+        "Pass_Secret": "usermanagerlvl48",
+        "Pass_SHA2": "9cc1ee455a3363ffc504f40006f70d0c8276648a5d3eb3f9524e94d1b7a83aef",
```
Comment (crnopster) [inline config/config.json]: Its not a good idea to push files like this

Comment (oklymchuk) [reply config/config.json]: Malfunctioned .gitignore. Fixed

File: config/config.go (hunk 1 above)
Comment (oklymchuk) [reply config/config.go]: Fixed,  thanks for prompt

File: config/config.go [hunk 3]
```diff
@@ -2,3 +2,46 @@
 // Basic configuration like consul credentials and address, http port to listen for requests,
 // postgres schema name, credentials, and client timeout are read from environment variables.
 package config
+
+import (
+	"strings"
+
+	"github.com/lvl484/user-manager/logger"
+	"github.com/spf13/viper"
```
Comment (i3odja) [inline config/config.go]: Maybe, will be better to insert empty line between "github.com/lvl484/user-manager/logger" and "github.com/spf13/viper". What do you think about it?

File: logger/umlogger.go [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+		log.SetFormatter(new(NullFormatter))
+	default:
+		return fmt.Errorf("Error logger configure output destination <%v> < should be Graylog, Stdout or File", lc.Output)
+	}
+	log.SetLevel(log.PanicLevel)
+	return nil
+}
+
+// Message log message depending to log level
+func Message(level log.Level, m string) {
```
Comment (radikh) [inline logger/umlogger.go]: Is this wrapper really needed?

File: logger/umlogger.go [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+		graylog_adr := fmt.Sprintf("%v:%v", lc.Host, lc.Port)
+		hook := graylog.NewGraylogHook(graylog_adr, map[string]interface{}{"API": "User management service"})
+		log.AddHook(hook)
+		log.SetFormatter(new(NullFormatter))
+	default:
+		return fmt.Errorf("Error logger configure output destination <%v> < should be Graylog, Stdout or File", lc.Output)
+	}
+	log.SetLevel(log.PanicLevel)
+	return nil
+}
```
Comment (radikh) [inline logger/umlogger.go]: You'd better to move "case" contents to a separate functions and just call them from here.

File: logger/umlogger.go [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+		log.SetOutput(os.Stdout)
+	case "File":
+		f, err := os.OpenFile("user_manager_api.log", os.O_APPEND|os.O_CREATE|os.O_RDWR, 0666)
+		if err != nil {
+			fmt.Printf("error opening file: %v", err)
+		}
+		log.SetFormatter(&log.JSONFormatter{})
+		log.SetOutput(f)
+	case "Graylog":
+		graylog_adr := fmt.Sprintf("%v:%v", lc.Host, lc.Port)
```
Comment (radikh) [inline logger/umlogger.go]: Go format uses camelCase for names, not snake_case.

File: config/config.go (hunk 3 above)
Comment (oklymchuk) [reply config/config.go]: Fixed

Comment (oklymchuk) [reply config/config.go]: fixed

File: logger/umlogger.go (hunk 4 above)
Comment (oklymchuk) [reply logger/umlogger.go]: Agry, we do not need it. Fixed

File: logger/umlogger.go (hunk 5 above)
Comment (oklymchuk) [reply logger/umlogger.go]: Fixed

File: logger/umlogger.go (hunk 6 above)
Comment (oklymchuk) [reply logger/umlogger.go]: Fixed

File: config/config.go [hunk 7]
```diff
@@ -2,3 +2,51 @@
...
+	loggerPassSecret = "loggerUM.PassSecret"
+	loggerPassSHA2   = "loggerUM.PassSHA2"
+	loggerOutput     = "loggerUM.Output"
+	LoggerLevel      = "loggerUM.Level"
+	LoggerType       = "loggerUM.Type"
+)
+
+type ViperCfg struct {
+	v *viper.Viper
+}
```
Comment (crnopster) [inline config/config.go]: Suggest to rename it to Config

File: config/config_test.go [hunk 8]
```diff
@@ -0,0 +1,102 @@
...
+			_, err := NewViperCfg(tt.args.configName, tt.args.configPath)
+			if (err != nil) != tt.wantErr {
+				t.Errorf("NewViperCfg() error = %v, wantErr %v", err, tt.wantErr)
+				return
+			}
+		})
+	}
+}
+
+func TestViperCfg_NewLoggerConfig(t *testing.T) {
```
Comment (crnopster) [inline config/config_test.go]: Use CamelCase instead of snake_case

File: config/config_test.go [hunk 9]
```diff
@@ -0,0 +1,102 @@
...
+				Host:       "GREYLOGHOST",
+				Port:       "77777",
+				PassSecret: "wrongpassword",
+				PassSHA2:   "anotherwrongpassword",
+				Output:     "Graylog",
+				Level:      "info",
+				Type:       "async",
+			},
+			wantErr: true,
+		},
```
Comment (crnopster) [inline config/config_test.go]: NewLoggerConfig does not return any errors

File: logger/umlogger.go [hunk 10]
```diff
@@ -0,0 +1,141 @@
...
+	} else {
+		hook = graylog.NewGraylogHook(graylogAdr, map[string]interface{}{"API": "User management service"})
+	}
+	log.AddHook(hook)
+	log.SetFormatter(new(NullFormatter))
+}
+
+// LoggerHandler creates a new middleware
+func LoggerHandler() func(http.Handler) http.Handler {
+	return func(handler http.Handler) http.Handler {
```
Comment (radikh) [inline logger/umlogger.go]: Why do you need two-level middleware?

File: logger/umlogger_test.go [hunk 11]
```diff
@@ -0,0 +1,278 @@
...
+			args:    args{e: entry},
+			want:    []byte{},
+			wantErr: false,
+		}, {
+			name:    "JSON format",
+			n:       NullFormatter{},
+			args:    args{e: &log.Entry{Logger: logger, Data: entry1}},
+			want:    []byte{},
+			wantErr: false,
+		},
```
Comment (radikh) [inline logger/umlogger_test.go]: Want will never be anything but empty slice of bytes. You even don't need this param.

File: config/config_test.go [hunk 12]
```diff
@@ -0,0 +1,102 @@
+// Package config is responsible for loading user-manager application config.
+// Basic configuration like consul credentials and address, http port to listen for requests,
+// postgres schema name, credentials, and client timeout are read from environment variables.
```
Comment (radikh) [inline config/config_test.go]: This comment is redundant.

File: config/config.go [hunk 13]
```diff
@@ -2,3 +2,51 @@
...
+func (vpcfg *ViperCfg) NewLoggerConfig() (*logger.LogConfig, error) {
+	return &logger.LogConfig{
+		Host:       vpcfg.v.GetString(loggerHost),
+		Port:       vpcfg.v.GetString(loggerPort),
+		PassSecret: vpcfg.v.GetString(loggerPassSecret),
+		PassSHA2:   vpcfg.v.GetString(loggerPassSHA2),
+		Output:     vpcfg.v.GetString(loggerOutput),
+		Level:      vpcfg.v.GetString(LoggerLevel),
+		Type:       vpcfg.v.GetString(LoggerType),
+	}, nil
```
Comment (radikh) [inline config/config.go]: It has error in return bur does not have error to return.

File: logger/umlogger.go [hunk 14]
```diff
@@ -0,0 +1,141 @@
...
+	Port       string
+	PassSecret string
+	PassSHA2   string
+	Output     string
+	Level      string
+	Type       string
+}
+type loggerKeyType int
+
+const loggerKey loggerKeyType = iota
```
Comment (radikh) [inline logger/umlogger.go]: iota does not work out of block of constants, if you need to set zero to constant - just set zero.

File: logger/umlogger.go (hunk 10 above)
Comment (oklymchuk) [reply logger/umlogger.go]: As I read in this article [writing-middleware-in-golang](https://medium.com/@matryer/writing-middleware-in-golang-and-how-go-makes-it-so-much-fun-4375c1246e81) we can capture the logger object inside our functions without spoiling the http.Handler.Do you think it it does not deserve?
Also i can add something like this
func Adapt(h http.Handler, adapters ...Adapter) http.Handler {
  for _, adapter := range adapters {
    h = adapter(h)
  }
  return h
}
but I don't know do we need this?

File: logger/umlogger_test.go (hunk 11 above)
Comment (oklymchuk) [reply logger/umlogger_test.go]: And how I must to test this function?
t.Errorf("NullFormatter.Format() = %v, want %v", got,[]byte{})?

File: config/config_test.go [hunk 15]
```diff
@@ -0,0 +1,102 @@
...
+package config
+
+import (
+	"reflect"
```
Comment (radikh) [inline config/config_test.go]: I would suggest to use assert in tests instead of reflect.

File: logger/umlogger.go [hunk 16]
```diff
@@ -0,0 +1,141 @@
...
+	case "File":
+		err = lc.setLoggerToFile()
+		if err != nil {
+			return err
+		}
+	case "Graylog":
+		lc.setLoggerToGraylog()
+	default:
+		err = ErrFailedToConfigureLog
+		return err
```
Comment (radikh) [inline logger/umlogger.go]:
```suggestion
		return ErrFailedToConfigureLog
```

File: logger/umlogger.go [hunk 17]
```diff
@@ -0,0 +1,141 @@
...
+	log.SetFormatter(&log.JSONFormatter{})
+	log.SetOutput(f)
+	return err
+}
+
+// setLoggerToStdout initialize logger for writing to stdout
+func (lc *LogConfig) setLoggerToStdout() {
+	log.SetFormatter(&log.TextFormatter{})
+	log.SetOutput(os.Stdout)
+}
```
Comment (radikh) [inline logger/umlogger.go]: Why two different formats? Suggest to make them configurable.

File: logger/umlogger.go [hunk 18]
```diff
@@ -0,0 +1,141 @@
...
+// New returns a context that has a logrus logger
+func New(ctx context.Context) context.Context {
+	return context.WithValue(ctx, loggerKey, WithContext(ctx))
+}
+
+// WithContext returns a logrus logger from the context
+func WithContext(ctx context.Context) *log.Logger {
+	if ctx == nil {
+		return log.StandardLogger()
+	}
```
Comment (radikh) [inline logger/umlogger.go]: Caller should newer pass nil context

File: logger/umlogger.go [hunk 19]
```diff
@@ -0,0 +1,141 @@
...
+
+	if ctxLogger, ok := ctx.Value(loggerKey).(*log.Logger); ok {
+		return ctxLogger
+	}
+
```
Comment (radikh) [inline logger/umlogger.go]: Do not pass logger in context, use function parameter instead.

Comment (oklymchuk) [reply logger/umlogger.go]: fixed

File: logger/umlogger.go (hunk 18 above)
Comment (oklymchuk) [reply logger/umlogger.go]: Removed

File: logger/umlogger.go (hunk 19 above)
Comment (oklymchuk) [reply logger/umlogger.go]: removed

File: logger/umlogger.go (hunk 17 above)
Comment (oklymchuk) [reply logger/umlogger.go]: left json format

File: logger/umlogger.go (hunk 16 above)
Comment (oklymchuk) [reply logger/umlogger.go]: fixed

File: config/config_test.go (hunk 15 above)
Comment (oklymchuk) [reply config/config_test.go]: Ok, used it

Comment (oklymchuk) [reply config/config_test.go]: But then do not need it

File: logger/umlogger.go (hunk 14 above)
Comment (oklymchuk) [reply logger/umlogger.go]: fixed, removing it

File: config/config.go (hunk 13 above)
Comment (oklymchuk) [reply config/config.go]: fixed

File: config/config_test.go (hunk 12 above)
Comment (oklymchuk) [reply config/config_test.go]: Made shorter

File: logger/umlogger_test.go (hunk 11 above)
Comment (oklymchuk) [reply logger/umlogger_test.go]: fixed

File: logger/umlogger.go (hunk 10 above)
Comment (oklymchuk) [reply logger/umlogger.go]: Removed, do not need

File: config/config_test.go (hunk 9 above)
Comment (oklymchuk) [reply config/config_test.go]: Is it bad, that it does not have error )))) Fixed

File: config/config_test.go (hunk 8 above)
Comment (oklymchuk) [reply config/config_test.go]: fixed

File: config/config.go (hunk 7 above)
Comment (oklymchuk) [reply config/config.go]: fixed

File: Dockerfile [hunk 20]
```diff
@@ -1,15 +1,16 @@
...
-ADD go.mod /m/
-RUN cd /m && go mod download && mkdir -p /go/pkg/
-
 FROM golang:1.13 as builder
-COPY --from=modules /go/pkg/ /go/pkg
-RUN mkdir -p /opt/resource/ 
-WORKDIR /opt/resource/
+RUN mkdir -p /go/src/github.com/lvl484
+ENV GO111MODULE on
+ENV CGO_ENABLED 0
```
Comment (alpaka-alpachino) [inline Dockerfile]: why you need that?

**Pull request merged**